runtime/src/accounts_db.rs

```diff
@@ -3110,9 +3116,13 @@ impl AccountsDB {
         // If `should_clean` is None, then`should_flush_f` is also None, which will cause
         // `flush_slot_cache` to flush all accounts to storage without cleaning any accounts.
         let mut should_flush_f = should_clean.map(|(account_bytes_saved, num_accounts_saved)| {
-            move |&pubkey: &Pubkey, account: &Account| {
+            move |slot: &Slot, &pubkey: &Pubkey, account: &Account| {
                 use std::collections::hash_map::Entry::{Occupied, Vacant};
-
+                if let Some(max_clean_root) = max_clean_root {
+                    if *slot > max_clean_root {
```

Comment:
@ryoqun Alternatively instead of checking this here on every account, I could move the check to here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L3136-L3139 outside of the `should_flush_f`, and pass `None` for `should_flush_f` when `root > max_clean_root` but the logic would be more spread out and harder to track. What do you think?

Reply:
As a conclusion, I don't have strong opinion on this. But there is my personal idea:

I understand your concern that logics are being spread. At the same time, I'm also worried that there could be many things in one place. This is hard call.

But separating might be more preferable in this case:
- `slot > max_clean_root` check here is sort of belong to the looping construct over `cached_roots` conceptually, suggesting the below `for` is more proper home for it. That's because given `slot` is always same inside a single iteration.
- if we move this check out of this closure, we can define `max_clean_root` immediately before the `for` loop, reducing being more spread out.
- Although cpu branch prediction/cpu cache nicely work in this case, it's best to hoist redundant code outside inner hot loop. If we urged to remain this code as-is, I think one-line note is needed not to confuse future readers of this code.

Reply:
added it to the `AccountsBackgroundService` closer to the snapshotting logic